mountpoint-s3/src/prefetch/backpressure_controller.rs

```diff
@@ -91,58 +115,125 @@ impl BackpressureController {
             let new_read_window_size =
-                (self.preferred_read_window_size * self.read_window_size_multiplier).min(self.max_read_window_size);
-            trace!(
-                current_size = self.preferred_read_window_size,
-                new_size = new_read_window_size,
-                "scaling up preferred read window"
+                align(new_read_window_size, self.read_part_size, false).min(self.max_read_window_size);
+
+            // Only scale up when there is enough memory. We don't have to reserve the memory here
+            // because only `preferred_read_window_size` is increased but the actual read window will
+            // be updated later on `DataRead` event (where we do reserve memory).
+            let to_increase = (new_read_window_size - self.preferred_read_window_size) as u64;
+            let available_mem = self.mem_limiter.available_mem();
+            if available_mem >= to_increase {
+                let formatter = make_format(humansize::BINARY);
+                debug!(
+                    current_size = formatter(self.preferred_read_window_size),
+                    new_size = formatter(new_read_window_size),
+                    "scaled up preferred read window"
+                );
+                self.preferred_read_window_size = new_read_window_size;
+                metrics::histogram!("prefetch.window_after_increase_mib")
+                    .record((self.preferred_read_window_size / 1024 / 1024) as f64);
+            }
+        }
+    }
+
+    // Scale down preferred read window size by a multiplier configured at initialization.
+    fn scale_down(&mut self) {
+        if self.preferred_read_window_size > self.min_read_window_size {
+            assert!(self.read_window_size_multiplier > 1);
+            let new_read_window_size = self.preferred_read_window_size / self.read_window_size_multiplier;
+            // Also align the new read window size to the client part size
+            let new_read_window_size =
+                align(new_read_window_size, self.read_part_size, false).max(self.min_read_window_size);
+
+            let formatter = make_format(humansize::BINARY);
+            debug!(
+                current_size = formatter(self.preferred_read_window_size),
+                new_size = formatter(new_read_window_size),
+                "scaling down read window"
             );
             self.preferred_read_window_size = new_read_window_size;
+            metrics::histogram!("prefetch.window_after_decrease_mib")
+                .record((self.preferred_read_window_size / 1024 / 1024) as f64);
         }
     }
 }
 
+impl<Client: ObjectClient> Drop for BackpressureController<Client> {
+    fn drop(&mut self) {
+        // Free up memory we have reserved for the read window.
+        debug_assert!(self.request_end_offset >= self.next_read_offset);
+        let remaining_window = self.read_window_end_offset.saturating_sub(self.next_read_offset);
```

Comment:
`push_front` and read would happen in the same fuse request because that's what trigger backwards seek, so that should not happen. But if we're worried then we can exclude the used mem in the backwards window.

Reply:
Sure, I see that it works today, but looks a bit fragile and I'd propose to have a mitigation here. 

As a matter of fact, if we [had pushed to the queue](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3/src/prefetch.rs#L525) before deciding if the seek was successful, we'd be reserving without ever releasing.

The mitigation could be as I proposed (not accounting data from `PartQueue::front_queue`) or something else if it's more convenient.

Reply:
The reason I wanted to push back is that we then have to decide whether the read is backwards seek at `BackpressureFeedbackEvent::DataRead` and if we don't be careful we could face the other side of the problem i.e. releasing memory more than we should and causing overflow which is also fragile in my opinion.